src/vs/vscode.proposed.d.ts

```diff
@@ -33,7 +33,38 @@ declare module 'vscode' {
 		extensionHostEnv?: { [key: string]: string | null };
 	}
 
-	export type ResolverResult = ResolvedAuthority & ResolvedOptions;
+	export interface Port extends Disposable {
+		remotePort: number;
+		localPort?: number;
+		description?: string;
+		forwardMechanism?: string;
+	}
+
+	/**
+	 * Used as part of the ResolverResult if the extension has any candidate, published, or forwarded ports.
+	 */
+	export interface PortInformation {
+		/**
+		 * Ports the extension knows are interesting and might be nice to forward.
+		 */
+		candidates?: Port[];
+		/**
+		 * Ports that are already immutably published. This is not the same as forwarding.
+		 */
+		published?: Port[];
+		/**
+		 * An array of ports to be forwarded once connected. For example,
+		 * if the extension wants to have a configuration file with some default ports, they could be passed through for forwarding here.
+		 */
+		toForward?: { remote: number, local: number, name: string, closeable?: boolean }[];
```

Comment:
What is the difference between `toForward` and `candidates` ? Why is here a type inlined instead of using `Port`. Perhaps `Port` should have the `closeable` property too.

Reply:
`candidates` are ports that the user might want to forward. We will show them in the port UI to make it easy/obvious that these might be useful ports. `toForward` are ports that the extension knows should definitely be forwarded. `toForward` ports will be forwarded.

Reply:
I don't think that `closeable` or `name` is appropriate in the other places. A `candidate` port doesn't need a name, because the user hasn't taken action on it yet. Same for a `published` port. Similarly, `candidate` and `published` ports aren't closeable. I'm also not sure that when an extension implements `forwardPort` they should be allowed to decide if that port that they're forwarding is `closeable`.

Reply:
But then `toForward` should be removed and the extension can call in a loop `asExternalUri`.

Reply:
`asExternalUri` doesn't allow the extension to give the port a name or choose a specific port.

Reply:
But then should we just add separate port forwarding API, i.e. acknowledge that `asExternalUri` is not sufficient and add something more powerful?

Reply:
I think the use of `asExternalUri` is different from `toForward`:
- Any extension can call `asExternalUri` to get a usable uri. Generally, I expect that extension calling `asExternalUri` only care that they have a usable uri, not the specifics of how they get that usable uri. We don't actually want them to care about specifics. They should "just work"
- Only a resolver can use `toForward` since it is used as part of the remote session setup. The session setup knows that it is remote, so it's the resolver's job to care about specifics such as port numbers.

Reply:
Given the functionality desire that a resolver can forward ports, there are two ways to go about it:

1. Add `PortInformation.toForward` as is proposed here. This is an expansion of the current API. This means that only resolvers can forward ports, but then they cannot stop the forwarding because `toForward` is provided by them once and then resolvers cannot call `Dispose` on something to stop them. So if a config file changes, it is not possible for a resolver to react to it and stop forwarding the current ports and begin forwarding the new ports.

2. Add `vscode.workspace.forwardPort(...): Port`. This means we do not add `PortInformation.toForward`. This is also an expansion of the current API. But this means any extension, including resolvers, can forward ports. Extensions end up forwarding ports today only as a side-effect of calling `asExternalUri`, so extensions already have the "access" to do port forwarding. Since `Port extends Disposable`, an extension could stop a certain port forwarding by calling `Dispose`. So if a config file changes, a resolver could potentially react to it and stop forwarding the current ports and begin forwarding the new ports.

I don't agree with the argument that `PortInformation.toForward` is somehow leading to a smaller or less expanded API compared to adding `vscode.workspace.forwardPort(...): Port`. IMHO they are both API expansions, with different characteristics.

---

Finally, I think `candidates: Port[]` is typed in a funny way in this proposal. A resolver would need to return `Port` objects as candidates, but what does it mean to call `dispose()` on such candidates?

Reply:
just FYI, today we let the user to choose port number and the name they desire, and they can dispose the port anytime they want.

Reply:
@fara-nak thanks for taking a look! Choosing a name and port number will be exposed through the UI.